[MiniBrowser] Automatically infer URL scheme if not provided

zakariaridouh opened the pull request:
#### 6e811aba9f4af5f5e3a31bc48703ecc2342c220a
<pre>
[MiniBrowser] Automatically infer URL scheme if not provided
<a href="https://bugs.webkit.org/show_bug.cgi?id=300840">https://bugs.webkit.org/show_bug.cgi?id=300840</a>
<a href="https://rdar.apple.com/162725335">rdar://162725335</a>

Reviewed by Elliott Williams.

Add capability to automatically build the URL scheme if not provided to `run-minibrowser`.

For example, run-minibrowser --url webkit.org doesn&apos;t work. Make it so we can infer the URL scheme and use it.

webkit.org -&gt; &apos;<a href="https://webkit.org&apos">https://webkit.org&apos</a>;
localhost:8080 -&gt; &apos;<a href="http://localhost">http://localhost</a>:8080&apos;
and so on.

* Tools/Scripts/webkitpy/minibrowser/run_webkit_app.py:
(main):

Canonical link: <a href="https://commits.webkit.org/301651@main">https://commits.webkit.org/301651@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/da27a5ab4b2fc836fa3e826a0b1ddae8ed0e5449

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126639 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46283 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37245 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/133608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/78299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/bd4b3d86-d310-4fe9-a411-4a999032cb87) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128510 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/46916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54821 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/133608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/78299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/13f81fe3-ca5c-4fc7-9de2-56fed9a61451) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129587 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/46916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/113266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/133608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7cf81e1f-adc2-49e5-a6e8-7d1d28936b1d) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/125991 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/46916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/31449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/77004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/46916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/31744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/136178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/53326 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/41026 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/136178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53815 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/109621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/136178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26671 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/50072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/28410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50758 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53248 "Built successfully") | | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/52525 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/55861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/54262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/Scripts/webkitpy/minibrowser/run_webkit_app.py:
```diff
@@ -54,12 +54,32 @@ def main(argv):
     if not options.platform:
         options.platform = "mac"
 
+    from urllib.parse import urlparse
```
style nit: move this to the top with the other imports

emw-apple commented on Tools/Scripts/webkitpy/minibrowser/run_webkit_app.py:
```diff
@@ -54,12 +54,32 @@ def main(argv):
+    def normalize_url(url):
+        """Normalize URL by adding https:// scheme if missing."""
+        if not url:
+            return url
```
This seems like a redundant check based on how this function is called.

emw-apple commented on Tools/Scripts/webkitpy/minibrowser/run_webkit_app.py:
```diff
@@ -54,12 +54,32 @@ def main(argv):
+        # If no scheme is given, add https://
+        if not parsed.scheme:
+            # Handle localhost and IP addresses specially
+            if url.startswith('localhost') or url.split(':')[0].replace('.', '').isdigit():
```
This won't handle IPv6 addresses, do we care? You could consider using the [stdlib `ipaddress` module](https://docs.python.org/3/library/ipaddress.html#ipaddress.ip_address) to parse this instead, but maybe that's overkill.

zakariaridouh replied:
Good point, I can followup with IPV6 support. I'll brainstorm on how best to do it.

emw-apple commented on Tools/Scripts/webkitpy/minibrowser/run_webkit_app.py:
```diff
@@ -54,12 +54,32 @@ def main(argv):
 
+    from urllib.parse import urlparse
+
+    def normalize_url(url):
```
style nit: I'd prefer that you inline this function into the one place it's used. You'll save a few lines of boilerplate code, and I don't think this is an important enough unit to maintain in isolation.

zakariaridouh replied:
That's fair, fixed in latest iteration!

emw-apple reviewed (approved):
Makes sense, here's a couple ideas for improvement.

zakariaridouh committed:
[MiniBrowser] Automatically infer URL scheme if not provided
https://bugs.webkit.org/show_bug.cgi?id=300840
rdar://162725335

Reviewed by Elliott Williams.

Add capability to automatically build the URL scheme if not provided to `run-minibrowser`.

For example, run-minibrowser --url webkit.org doesn't work. Make it so we can infer the URL scheme and use it.

webkit.org -> 'https://webkit.org'
localhost:8080 -> 'http://localhost:8080'
and so on.

* Tools/Scripts/webkitpy/minibrowser/run_webkit_app.py:
(main):

Canonical link: https://commits.webkit.org/301651@main
